docs/reference/configuration/configuration.md

```diff
@@ -1,48 +1,276 @@
+
+```JSON
+"target_overrides": {
+    "*": {
+        "cellular.random_max_start_delay": "100"
+    },
+    "K64F": {
+        "cellular.use-apn-lookup": false
+    }
+}
+```
+
+Examining the configuration for the target `LPC1768` with `mbed compile --config -m LPC1768` results in the following configuration:
+
+```
+Configuration parameters
+------------------------
+cellular.random_max_start_delay = 100 (macro name: "MBED_CONF_CELLULAR_RANDOM_MAX_START_DELAY")
+cellular.use-apn-lookup = 1 (macro name: "MBED_CONF_CELLULAR_USE_APN_LOOKUP")
+<output truncated for brevity>
+```
+
+Examining the configuration for the target `K64F` with `mbed compile --config -m K64F` results in the following configuration:
+
+```
+Configuration parameters
+------------------------
+cellular.random_max_start_delay = 100 (macro name: "MBED_CONF_CELLULAR_RANDOM_MAX_START_DELAY")
+cellular.use-apn-lookup = 0 (macro name: "MBED_CONF_CELLULAR_USE_APN_LOOKUP")
+<output truncated for brevity>
+```
+
+The order in which overrides are considered is:
+
+ 1. Libraries override target configuration with `mbed_lib.json`.
+ 2. The application overrides target and library configuration with `mbed_app.json`
+
+#### Defining configuration parameters
+
+The configuration system understands configuration parameters that targets, libraries and applications define. A JSON object called "config" always defines the configuration parameters.
+
+For example:
 
 ```JSON
 {
     "config": {
-        "main-stack-size": {
-            "value": 2000
-        }
+        "param1": {
+            "help": "The first configuration parameter",
+            "macro_name": "CUSTOM_MACRO_NAME",
+            "value": 0
+        },
+        "param2": {
+            "help": "The second configuration parameter",
+            "required": true
+        },
+        "param3": 10
     }
 }
 ```
 
-#### Configuring user spawned thread stacks
+You define a configuration parameter by specifying its name as the key and specifying its value either with a description object or by value. The JSON fragment above defines three configuration parameters named `param1`, `param2` and `param3`.
+
+Above, the configuration parameters `param1` and `param2` are defined using a description object. The following keys are supported on the description object:
+
+  - `help`: an optional help message that describes the purpose of the parameter.
+  - `value`: an optional field that defines the value of the parameter.
+  - `required`: an optional key that specifies whether the parameter must be given a value before compiling the code (`false` by default). It's not possible to compile a source tree with one or more required parameters that don't have a value. Generally, setting `required` to true is only useful when `value` is not set.
```

Comment:
Query: Who or what gives the parameter a value before compiling the code? You, the user? Please clarify for active voice.

Reply:
how about must have a value?